hw02_unpack_string/unpack.go

```diff
@@ -2,11 +2,112 @@ package hw02unpackstring
 
 import (
 	"errors"
+	"strconv"
+	"strings"
+	"unicode"
 )
 
 var ErrInvalidString = errors.New("invalid string")
 
-func Unpack(_ string) (string, error) {
-	// Place your code here.
-	return "", nil
+func unpackInitState(s string) ([]rune, int, strings.Builder) {
+	runes := []rune(s)
+	runesLength := len(runes)
+	var result strings.Builder
+	return runes, runesLength, result
+}
+
+func unpackValidator(runesLength int, runes []rune) (bool, error) {
+	if runesLength == 0 {
+		return true, nil
+	}
+	if unicode.IsDigit(runes[0]) {
+		return true, ErrInvalidString
+	}
+	return false, nil
+}
+
+func initIterationState(runes []rune, idx int) (rune, int, int, bool) {
+	currentRune := runes[idx]
+	nextIdx := idx + 1
+	idxAfterNext := nextIdx + 1
+	currentRuneIsSlash := string(currentRune) == "\\"
+	return currentRune, nextIdx, idxAfterNext, currentRuneIsSlash
+}
+
+func writeLetter(r rune, builder *strings.Builder) {
+	if unicode.IsLetter(r) {
+		builder.WriteRune(r)
+	}
+}
+
+func writeSeq(ch string, multStr string, builder *strings.Builder) {
+	mult, _ := strconv.Atoi(multStr)
+	builder.WriteString(strings.Repeat(ch, mult))
+}
+
+func isSlash(str string) bool {
+	return str == "\\"
+}
+
+func isDigit(r rune) bool {
+	return unicode.IsDigit(r)
+}
+
+func getUnpackedString(builder *strings.Builder, err error) (string, error) {
+	return builder.String(), err
+}
+
+func Unpack(s string) (string, error) {
+	runes, runesLength, result := unpackInitState(s)
+	invalidStr, err := unpackValidator(runesLength, runes)
```

Comment:
валидации лучше прописывать сразу в начале функции unpack
```suggestion
        invalidStr, err := unpackValidator(len(string), runes)
        if err != nil {return err}
	runes, result := unpackInitState(s)
	
```